architecture.md

```diff
@@ -0,0 +1,84 @@
+# Shepherd Architecture
+
+This document defines the architecture of Shepherd and the technologies used.
+
+## Overall
+
+Shepherd will consist of the following services:
+
+- A web server to receive queries from the ARS, developers, users, etc.
+- A relational database to store query state/metadata
+- A JSON object store for query and response objects
+- A message broker to distribute tasks to all shepherd services
+- Operation workers to perform all Translator ARA functions
+
+![Shepherd Architecture](./Shepherd%20Architecture.jpg)
+
+### Web Server
+
+The web server will use FastAPI:
+
+- It is written in Python, so the majority of Translator developers will be able to understand and interact with it
+- It has native support for OpenAPI and thus be able to be used in SmartAPI
+- It has great support for OpenTelemetry. With a few extra imports and lines of code, it will be fully integrated with translator-otel.
+- It is very performant and can handle 1000s of concurrent requests
+- It can scale easily in Kubernetes by spinning up new instances
+
+### Relational Database
+
+The relational database will use PostgreSQL:
+
+- It is a tried and true relational database
+- It uses SQL, which is a standard language for database interactions
+- An [ARAX tool](https://arax.ncats.io/?recent=24) already exists that can be used for monitoring the database
+- There are ways to horizontally scale it by sharding, replication, etc.
+
+#### Database Schema
+
+There will be two database tables:
+
+- shepherd_brain:
+  - parent_pk: Given by the ARS, is the uuid of an initial query
+  - child_pk: Given by the ARS, is the uuid of a query sent to a specific ARA
+  - start_time: The timestamp in UTC of when the query was first received
+  - stop_time: The timestamp in UTC of when the query was completed
+  - submitter: The name of the service that sent the query
+  - remote_ip: The IP address of the submitter
+  - domain: The url of the receiving Shepherd server
+  - hostname: The uuid of the receiving Shepherd server process
+  - response_id: The uuid of the response message of the query
+  - callback_url: For async queries, given in the initial query, where the response should be POSTed back once completed
+  - state: The current state of the query, i.e. Queued, Running, Completed, etc.
+  - status: The current status of the query, i.e. Ok, No Results, other errors, etc.
+  - description: A short text description containing query and/or error details
+- callbacks:
+  - child_pk: A foreign key, links back to the child_pk that is stored in shepherd_brain
```

Comment:
What is meant by a "foreign key"?

Reply:
https://en.wikipedia.org/wiki/Foreign_key Basically, it would be guaranteed that the child_pk in the callbacks table is also in the shepherd_brain table.